weblab/entities/views.py

```diff
@@ -239,33 +242,63 @@ def get_context_data(self, **kwargs):
         kwargs['delete_file'] = self.request.GET.get('deletefile')
         return super().get_context_data(**kwargs)
 
-    def post(self, request, *args, **kwargs):
-        entity = self.get_object()
+    def add_and_delete_files(self, deletions, additions):
+        pass
 
-        if 'delete_filename[]' in request.POST:
-            deletions = request.POST.getlist('delete_filename[]')
-            for filename in deletions:
-                path = str(entity.repo_abs_path / filename)
-                entity.delete_file_from_repo(path)
+    def post(self, request, *args, **kwargs):
+        entity = self.object = self.get_object()
 
-        if 'filename[]' in request.POST:
-            additions = request.POST.getlist('filename[]')
-            uploads = entity.files.filter(upload__in=additions)
+        git_errors = []
+        files_to_delete = []
 
-            # Copy each file into the git repo
-            for upload in uploads:
-                src = os.path.join(settings.MEDIA_ROOT, upload.upload.name)
-                dest = str(entity.repo_abs_path / upload.original_name)
-                shutil.move(src, dest)
+        # Delete files from the index
+        deletions = request.POST.getlist('delete_filename[]')
+        for filename in deletions:
+            path = str(entity.repo_abs_path / filename)
+            try:
+                entity.delete_file_from_repo(path)
+            except GitCommandError as e:
+                git_errors.append(e.stderr)
+
+        # Copy files into the index
+        additions = request.POST.getlist('filename[]')
+        for upload in entity.files.filter(upload__in=additions):
+            src = os.path.join(settings.MEDIA_ROOT, upload.upload.name)
+            dest = str(entity.repo_abs_path / upload.original_name)
+            files_to_delete.append(src)
+            shutil.copy(src, dest)
+            try:
                 entity.add_file_to_repo(dest)
-
-        entity.commit_repo(request.POST['commit_message'],
-                           request.user.full_name,
-                           request.user.email)
-        entity.tag_repo(request.POST['version'])
-
-        return HttpResponseRedirect(
-            reverse('entities:%s' % entity.entity_type, args=[entity.id]))
+            except GitCommandError as e:
+                git_errors.append(e.stderr)
+
+        if git_errors:
+            # If there were any errors with adding or deleting files,
```

Comment:
Will Django do something sensible with the temporary upload files in this case, keeping them for a resubmission attempt?

Also, what happens if the user abandons the new version? Does Django clean up automatically after an interval?

Reply:
@helenst can you answer the above queries?

Reply:
The temporary upload files will stay on the system (they're associated with the entity upon upload). There isn't a clean up yet, but probably should be. I'm not sure exactly at what point to decide it's appropriate to clean up an upload, since abandoning a version isn't a positive action by the user. Nightly job that cleans up anything older than 24 hours perhaps?

Reply:
That sounds sensible.